cmd/snap-bootstrap/cmd_scan_disk.go

```diff
@@ -0,0 +1,290 @@
+	fallback := false
+	var fallbackPartition string
+
+	bootUUID, err := bootFindPartitionUUIDForBootedKernelDisk()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "No efi var, falling back: %s\n", err)
+		fallback = true
+	}
+
+	partitions, err := probePartitions(node)
+	if err != nil {
+		return fmt.Errorf("Cannot get partitions: %s\n", err)
+	}
+
+	/*
+	 * If LoaderDevicePartUUID was not set, it is probably because
+	 * we did not boot with UEFI. In that case we try to detect
+	 * disk with partition labels.
+	 */
+	if fallback {
+		mode, _, err := boot.ModeAndRecoverySystemFromKernelCommandLine()
+		if err != nil {
+			return err
+		}
+		switch mode {
+		case "recover":
+			fallbackPartition = "ubuntu-seed"
+		case "install":
+			fallbackPartition = "ubuntu-seed"
+		case "factory-reset":
+			fallbackPartition = "ubuntu-seed"
+		case "run":
+			fallbackPartition = "ubuntu-boot"
+		case "cloudimg-rootfs":
+			fallbackPartition = "ubuntu-boot"
+		default:
+			return fmt.Errorf("internal error: mode not handled")
+		}
+	}
+
+	/*
+	 * If we are not in UEFI mode and snapd_system_disk is
+	 * defined, we need to verify the disk also matches that. If
+	 * not, we just return, ignoring this disk.
+	 */
+	if fallback {
+		values, err := kcmdline.KeyValues("snapd_system_disk")
+		if err != nil {
+			return fmt.Errorf("Cannot read kernel command line: %s\n", err)
+		}
+
+		if value, ok := values["snapd_system_disk"]; ok {
+			var currentPath string
+			var expectedPath string
+			if strings.HasPrefix(value, "/dev/") || !strings.HasPrefix(value, "/") {
+				name := strings.TrimPrefix(value, "/dev/")
+				expectedPath = fmt.Sprintf("/dev/%s", name)
+				currentPath = node
+			} else {
+				expectedPath = value
+				currentPath = osGetenv("DEVPATH")
+			}
+
+			same, err := samePath(filepath.Join(dirs.GlobalRootDir, expectedPath),
+				filepath.Join(dirs.GlobalRootDir, currentPath))
+			if err != nil {
+				return fmt.Errorf("Cannot check snapd_system_disk kernel parameter: %s\n", err)
+			}
+			if !same {
+				return nil
+			}
+		}
+	}
+
+	/*
+	 * Now we scan the partitions. For each partition we need to find out:
+	 *  - If its label matches a known partition type. In this
+	 *    case we save the UUID of that partition.
+	 *  - If the paritition is the booted ESP. In that case, we can confirm
+	 *    we are looking at the booted disk.
+	 */
+	found := false
+	hasFallbackPartition := false
+	hasSeed := false
+	hasBoot := false
+	for _, part := range partitions {
+		if !fallback {
+			if part.UUID == bootUUID {
+				/*
+				 * We have just found the ESP boot partition!
+				 */
+				found = true
+			}
+		}
+		if fallback && part.Name == fallbackPartition {
+			/*
+			 * We are not in UEFI boot, and we have found
+			 * a partition that looks like the boot
+			 * partition.
+			 */
+			hasFallbackPartition = true
+		}
+		if part.Name == "ubuntu-seed" {
+			hasSeed = true
+		} else if part.Name == "ubuntu-boot" {
+			hasBoot = true
```

Comment:
Not sure if looking for the boot partition is worth it, the one that always is present is the seed. And the check `hadSeed || hasBoot` would pass if we have a boot part. but not a seed part., which is certainly a worrying case where something is very broken. But detecting it is probably a nice thing to print to `os.Stderr` I guess.

Reply:
We need in case we are in non-UEFI mode.

Reply:
Right. I tried to express exactly what snap-bootstrap has. And if we have corner cases with a boot partition only. That said, it is probably not possible with 24+.